man/makeplist.8
.Dd 23 February, 2021
.Dt MAKEPLIST 8
.Os
.Sh NAME
.Nm makeplist
.Nd generate a
.Li pkg-plist
file for
.Xr ports 7
.Sh SYNOPSIS
.Nm
.Op Fl lOq
.Op Fl I Ar make-vars
.Op Fl o Ar plist-file
.Op Ar port-name
.Nm
.Fl h
.Sh DESCRIPTION
The
.Nm
script is a tool for
.Xr ports 7
maintainers and committers, to generate a
.Pa pkg-plist
file.
.Ss Arguments
The following arguments are accepted:
.Bl -tag -width indent
.It Ar make-vars
A comma separated list of make variables. The following variables
may be specified:
.Pp
.Bl -bullet -compact
.It
.Li DESKTOP_ENTRIES
.It
.Li PLIST_FILES
.It
.Li PLIST_SUB
.It
.Li PORTDATA
.It
.Li PORTDOCS
.It
.Li PORTEXAMPLES
.It
.Li USE_RC_SUBR
.El
.It Ar plist-file
The output file for the generated plist. Defaults to the ports
.Pa pkg-plist
with a
.Pa .makeplist
suffix appended.
.It Ar port-name
The name or origin of the port to generate a pkg-plist for. This may
also be the directory the port can be found in.
.Pp
If omitted the current directory is used.
.El
.Ss Options
The following options are support:
.Bl -tag -width indent
.It Fl h , -help
Displays the available options.
.It Fl I Ar make-vars , Fl -ignore-vars Ar make-vars
Set the make variables to ignore when postprocessing the generated plist.
.It Fl l , -licenses
Usually
.Nm
runs with with the
.Xr ports 7
license auditing framework disabled. This option turns it on.
.Pp
This flag cannot be combined with
.Fl q .
.It Fl O , -orig
Includes files with the .orig suffix in the plist.
.It Fl o Ar plist-file , Fl -outfile Ar plist-file
Sets the output file for the generated plist.
.It Fl q , -quiet
Suppress build/stage output.
.Pp
This flag cannot be combined with
.Fl l .
.El
.Sh IMPLEMENTATION NOTES
In order to make effective use of
.Nm
some knowledge about it is required.
.Ss Limitations
The exponential growth of possible combinations of options puts a
hard limit on automated plist generation. The
.Nm
script does not support cases where files are only installed if a
combination of options is given.
.Pp
The tool is built around the assumption that options do not affect
each other. The following subsections describe what exactly is supported
and workarounds for common cases where this is not the case.
.Ss Configurations
The core idea of
.Nm
is to run
.Dq Li make stage
and create a list of the files in the staging area. This functionality
is provided by the
.Xr ports 7
when running
.Dq Li make makeplist .
.Pp
In order to identify all the files installed by each option staging
is performed for every option. The trivial approach of staging with
one option at a time is not possible when
.Li OPTIONS_SINGLE
or
.Li OPTIONS_MULTI
groups are defined, because every such group must be represented in
all configurations. Even so under the assumption that options do not
affect each other n + 1 stage cycles suffice to identify all files
provided by all options (n being the number of options).
.Ss Staging
Usually
.Nm
calls
.Dq Li make clean stage
for each configuration. One exception to this rule is when ports define
.Li NO_BUILD
in which case
.Dq Li make restage
is called. This means that the
.Li extract
target is only performed during the first stage cycle. And it breaks
if extraction is affected by options.
.Ss Build Failure
In case one or more configurations fail an error message with the
build options and the name of the log file will be printed after
the plist is created. Only logs of failed builds are kept. The logs
are created under
.Pa /tmp
and compressed using
.Xr gzip 1 .
They can be viewed with the command
.Dq Li gunzip -c Ar logfile Li | less -R .
.Ss Plist Accumulation
After every stage cycle all files installed into the
.Li STAGEDIR
are collected for later assembly of the plist.
.Pp
Certain files are not included, these files are selected using the
following
.Xr ports 7
variables:
.Bl -tag -width indent
.It Li DESKTOP_ENTRIES
Macro for creating/installing desktop icons.
.It Li USE_RC_SUBR
Macro for creating/installing
.Xr rc 8
scripts.
.It Li PLIST_FILES
A list of files automatically added to the plist.
.It Li PORTDOCS , PORTEXAMPLES , PORTDATA
Lists of files / glob patterns in
.Li DOCSDIR , EXAMPLESDIR
and
.Li DATADIR .
.El
.Pp
This can be used to deal with cases that violate the independent
option principle. E.g. if there is a number of options to switch
certain modules on/off and a
.Li DOCS
option which causes each of these modules to generate some documentation,
setting
.Li PORTDOCS=*
keeps everything in
.Li DOCSDIR
out of the plist.
.Ss Plist Creation
Files in the generated plist are sorted alphabetically by
.Xr sort 1 Fl n .
.Pp
Common files installed independent of the given options are listed
first, followed by the option specific files. Option specific files
appear in alphabetical order of the options. The options
.Li DOCS
and
.Li EXAMPLES
are implicitly replaced with
.Li PORTDOCS
and
.Li PORTEXAMPLES .
Files with a
.Pa .sample
suffix are automatically prefixed with
.Dq Li @sample\  .
.Pp
In the next stage of plist creation the reverse of the substitutions
defined in
.Li PLIST_SUB
is applied. The substitutions are sorted by size and applied largest
first. Empty substitutions,
.Li PREFIX
substitutions and substitutions starting with an
.Li @
character are discarded. Lines starting with
.Li %%\&DOCSDIR%%
are prefixed with
.Li %%\&PORTDOCS%%
and
.Li %%\&EXAMPLESDIR%%
with
.Li %%\&PORTEXAMPLES%% .
.Pp
In the final stage of plist creation
.Nm
attempts to transplant the
.Li @
keywords (see
.Xr pkg-create 8 )
from the old plist to the new one. The script makes an effort to
place them in the same context. The prefix
.Dq Li @sample\ 
is stripped from all files that appear in the old plist without it.
.Sh EXAMPLES
The current directory is a port to create a pkg-plist for:
.Pp
.Dl % makeplist
.Pp
The quiet mode suppresses the build output:
.Pp
.Dl % makeplist -q
.Pp
Instead of creating the file
.Pa pkg-plist.makeplist
overwrite the
.Pa pkg-plist
file:
.Pp
.Dl % makeplist -o pkg-plist
.Pp
To create a pkg-plist as an unprivileged user run:
.Pp
.Dl % env WRKDIRPREFIX='/var/tmp/obj.${USER}' makeplist -o ~/myplist
.Pp
The
.Li WRKDIRPREFIX
can be set in the
.Xr make.conf 5 file:
.Pp
.Dl WRKDIRPREFIX?=/var/tmp/obj.${USER}
.Pp
There is no need to
.Li cd
into a ports directory:
.Pp
.Dl % makeplist games/ioquake3-devel
.Pp
The ports tree does not have to be in
.Pa /usr/ports :
.Pp
.Dl % env PORTSDIR=$HOME/ports.svn makeplist games/ioquake3-devel
.Pp
Or just use
.Li cd :
.Pp
.Dl % cd ~/ports.svn/games/ioquake3-devel && makeplist
.Pp
A port listing all its files in
.Li PLIST_FILES
results in an empty plist.
.Li PLIST_FILES
can be ignored:
.Pp
.Dl % makeplist -I PLIST_FILES
.Sh SEE ALSO
.Xr ports 7 , Xr make 1 , Xr pkg-greate 8
.Sh HISTORY
The
.Nm
script first appeared in the bsda2-0.2.0 release.
.Sh AUTHORS
.An Dominic Fandrey Aq Mt freebsd@k4m1.org


src/makeplist.awk
#!/usr/bin/awk -f
#
# Combine multiple pkg-plist files into a single pkg-plist.
#
# This takes a list of all options as command line arguments. The
# order of the options determines the order option specific plist
# entries appear in the combined output.
#
# The plists must be fed into stdin and preceded by a line starting
# with `OPTIONS:` followed by a white space separated list of the
# active options.
#
# Files that appear with all options are listed first, followed by
# the option specific output.
#
# @param ARGV
#	A sorted list of options
#

# Get the order of options
BEGIN {
	OPTION_STR["DOCS"] =     "%%PORTDOCS%%"
	OPTION_STR["EXAMPLES"] = "%%PORTEXAMPLES%%"
	for (i = 1; i < ARGC; ++i) {
		OPTIONS_ORDERD[i] = ARGV[i]
		CNT_OPT_FILES[ARGV[i]] = 0
		if (!(ARGV[i] in OPTION_STR)) {
			OPTION_STR[ARGV[i]] = "%%" ARGV[i] "%%"
		}
		if (!("NO_" ARGV[i] in OPTION_STR)) {
			OPTION_STR["NO_" ARGV[i]] = "%%NO_" ARGV[i] "%%"
		}
		delete ARGV[i]
	}
	CNT_FILES = 0
}

# Get the options the following files were staged with
/^OPTIONS:/ {
	delete aoptions
	sub(/^OPTIONS: */, "")
	cnt_aoptions = split($0, aoptions)
	for (i = 1; i <= cnt_aoptions; ++i) {
		++OPTIONS[aoptions[i]]
	}
	++CONFIGS
	next
}

# Collect files
{
	# Record order of file
	if (!($0 in FILES)) {
		FILES_ORDERED[++CNT_FILES] = $0
	}
	# Count occurence of file
	++FILES[$0]
	# The same book keepin per option
	for (i = 1; i <= cnt_aoptions; ++i) {
		option = aoptions[i]
		# Record order of file for option
		if (!OPT_FILES[option, $0]) {
			OPT_FILES_ORDERED[option, ++CNT_OPT_FILES[option]] = $0
		}
		# Count occurence of file by option
		++OPT_FILES[option, $0]
	}
}

# Print files
END {
	# Print files common to all configurations
	for (i = 1; i <= CNT_FILES; ++i) {
		file = FILES_ORDERED[i]
		if (FILES[file] == CONFIGS) {
			print file
			delete FILES[file]
		}
	}
	# Print option specific files
	for (i = 1; OPTIONS_ORDERD[i]; ++i) {
		option = OPTIONS_ORDERD[i]
		for (p = 1; p <= CNT_OPT_FILES[option]; ++p) {
			file = OPT_FILES_ORDERED[option, p]
			# Skip files that have already been printed
			if (!(file in FILES)) { continue }
			# Print file if it only occurs for the current
			# option
			if (OPT_FILES[option, file] == OPTIONS[option] &&
			    OPT_FILES[option, file] == FILES[file]) {
				print OPTION_STR[option] file
				delete FILES[file]
			}
		}
		for (p = 1; p <= CNT_FILES; ++p) {
			file = FILES_ORDERED[p]
			if (!(file in FILES)) { continue }
			# Print file if it occurs everywhere but with
			# this option
			if (!OPT_FILES[option, file] &&
			    FILES[file] + OPTIONS[option] == CONFIGS) {
				print OPTION_STR["NO_" option] file
				delete FILES[file]
			}
		}
	}
	# Print all files that have not been printed
	for (i = 1; i <= CNT_FILES; ++i) {
		file = FILES_ORDERED[i]
		if (!(file in FILES)) { continue }
		print "@fail " file " could not be mapped to an option!"
		msg = "@fail Candidates:"
		for (p = 1; OPTIONS_ORDERD[p]; ++p) {
			if (OPT_FILES[OPTIONS_ORDERD[p], file]) {
				msg = msg " " OPTIONS_ORDERD[p]
			}
		}
		print msg
	}
}
